File: May2022/C++/Count_Sorted_Vowel_Strings.cpp
class Solution {
public:
    int countVowelStrings(int n) {
        vector<vector<int>> dp(5,vector<int>(n+1,0));
        for(int i=0;i<5;i++) dp[i][1]=1;
        int sum=5,temp=5;
        for(int i=2;i<=n;i++)
        {
            sum=temp;
            temp=0;
            for(int j=0;j<5;j++)
            {
                if(j==0) dp[j][i]=sum;
                else
                {
                    sum-=dp[j-1][i-1];
                    dp[j][i]=sum;
                }
                temp+=dp[j][i];
            }
        }
        sum=temp;
        
        
        return sum;
        
        
    }
};

File: May2022/Java/CountSortedVowelStrings.java

// @saorav21994
// TC : O(5*n)
// SC : O(5*n)
// Cannot do plain recursion as it will give TLE

class Solution {
    public int countVowelStrings(int n) {
        int [][] dp = new int[n+1][6];
        dp[1][1] = 1;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j < 6; j++) {
                if (i == 1) {
                    dp[i][j] = dp[i][j-1] + 1;
                }
                else {
                    dp[i][j] = dp[i][j-1] + dp[i-1][j];
                }
            }
        }
        return dp[n][5];
    }
}


// Nice tutorial taken from GFG for constant time and space. Completely based on pattern observation.
// TC : O(1)
// SC : O(1)

    public int countVowelStrings(int n) {
        
        return (n+1)*(n+2)*(n+3)*(n+4)*1/24;
    
    }
